README.md
Introduction
------------

Hyde.el is an Emacs major mode to help create blogs with the excellent
[Jekyll](http://jekyllrb.com/) blogging system. It comes with a front
end (`hyde.el`), a version control backend (`hyde-git.el`) and an
slightly modified version of the stock markdown editing mode that
gives you a few bells and whistles while writing posts (`hyde-md.el`).


Installation
------------

### Hyde.el
Download the all the `hyde-*.el` files and put them somewhere. Once
you do that, add the directory where you put it to your load path like
so and `require` it.


    (add-to-list 'load-path "/path/to/hyde*.el")
    (require 'hyde)

### Jekyll

There are no hyde specific things you'll have to do for your Jekyll
installation. However, there is a `.hyde.el` file which you can drop
into your blog directory which contains blog specific settings. This
file will is not something that jekyll is aware of so you might want
to update your `_config.yml` to not process this file.


Operation
---------
This mode is a simply a wrapper for a number of shell commands that
are used to create and deploy the site. It doesn't maintain any local
state (in the form of status files etc.) so if you change your
repository manually outside it, just refreshing the buffer will bring
it upto date.

It's tailored to the way I work. I keep my posts in a `git` repository
(although I do have a crude DVCS abstraction layer if you're using
`hg` or any other such system). I make changes, commit them and push
the repository to `github` (you can, for example, see this files
source at
[github](https://github.com/nibrahim/nibrahim.net.in/blob/master/_posts/2010-11-11-hyde_%3A_an_emacs_mode_for_jekyll_blogs.markdown)).
After that, I make the site using `jekyll` manually and then copy it
over to my webspace using `rsync`. I don't use any of the git hooks
(yet).

Customisation
-------------
    
Following are the variables affecting your blog and its management that you can customise

  * `hyde-home` : The default root directory of your blog
  * `hyde-deploy-dir` : The directory where `jekyll` will generate the site for you to deploy
  * `hyde-posts-dir` : The directory that will contain the actual posts
	  (this is relative to `hyde-home` and is `_posts` by default).
  * `hyde-drafts-dir` : The directory that will contain the post
	  drafts (this is relative to `hyde-home` and is `_drafts` by
	  default. You might want to _ignore_ this directory in
	  `_config.yml` ).
  * `hyde/deploy-command` : The command used to deploy the site. `scp`,
	  `rsync` or whatever else you might please.
  * `hyde/jekyll-command` : The command used to run jekyll to generate the blog. You can add bits to take care of rvm for you here. 		  
  * `hyde/hyde-list-posts-command` : The command used to list the posts in the hyde buffer. It is set to `"/bin/ls -1tr "` by default which will produce a chronologically ordered list. You can change it if you prefer alphabetic or something else. 
  * `hyde/git/remote` : The remote to which the push command should send changes to. `origin` by default.
  * `hyde/git/remote-branch` : The branch on which you blog resides and to which the push will happen. `master` by default.

When you start hyde using `M-x hyde`, it will prompt you for the directory where your blog resides. All the paths mentioned above (`posts-dir` etc.) are all relative to this. The directory that you specify here will override `hyde-home` mentioned above. 

Customising any of the above will set them globally (i.e. for all blogs managed by jekyll on your system). If you have multiple blogs which you're managing using jekyll, you can drop a `.hyde.el` file into the blog directory where you can manually set any of these variables. In this file, you can manually change any of these variables using `setq`. There is a `sample-dot-hyde.el` file which shows you how. These settings override the global ones mentioned above and you can have different such settings for different blogs on your system. 

The following are commands and predicates used to handle the version control backend. You can change them if you want to add support for a version control system other than `git`. If it works for you, please send me a pull request and I'll integrate it. 

  * `hyde/vc-uncommittedp` : Predicate to check whether the file is uncommitted
  * `hyde/vc-unpushedp` : Predicate to check whether the file is not yet pushed
  * `hyde/vc-pushedp` :  Predicate to check if the file has been pushed (inverse of the above)
  * `hyde/vc-add` : Command to add the file
  * `hyde/vc-commit` : Command to commit a file
  * `hyde/vc-push` : Command to push the local changes to the remote end.
  * `hyde/vc-rename` : Command used to rename a file.
  
Interface
---------
  
The main interface looks like the following screenshot

![Hyde screenshot](Screenshot.png)

The list of posts are presented on top along with a key of what the
letters before the post names mean. The post names are also colourised
accordingly

The keys you can use at this time are

* `n` : Create a new draft.
* `c` : Commit the current post
* `P` : Push all pending commits (this is only a VC push. Not
  deployment).
* `j` : Run jekyll and create the new version of the site
* `d` : Deploy the site.
* `g` : Refresh posts (useful if you've done something by hand
  earlier)
* `p` : Promote a post from a draft to a a published post
* `q` : Quit hyde.  
* `RET` : Open the current post for editing.

The markdown mode in which the buffers open up for editing is slightly
modified. It has a few extra covenience bindings

* `C-c C-c` : Save file and finish editing
* `C-c C-v` : Preview file (this is a markdown preview so extra
  `liquid` tags will not work). 

License
-------
This program is licensed under the GNU General Public License Version
3. Please check the LICENSE file for the full text of the license.

To Do
-----

* Proper previews using a local Jekyll server
* Keep state of deployment so that we know what posts have been
  deployed and what not.
* Use `comint` instead of shell commands to do all the work. 


CHANGELOG
11 Dev 2011 - ver 0.2
 - Multiple blog support
 - Drafts support
 - Customisable remote and branch
 - Bumped up License to GPLv3

27 Dec 2010 - ver 0.1
- Initial release


hyde-md.el
;;; hyde-md.el
;; Copyright (C) 2004 Noufal Ibrahim <noufal at nibrahim.net.in>
;;
;; This program is not part of Gnu Emacs
;;
;; hyde-md.el is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published
;; by the Free Software Foundation; either version 3 of the License,
;; or (at your option) any later version.
;;
;; This program is distributed in the hope that it will be useful, but
;; WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
;; General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with this program; if not, write to the Free Software
;; Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
;; 02111-1307, USA.

(defun hyde-markdown-end-edit ()
  "Function called signifying the end of the editing session"
  (interactive)
  (save-buffer (current-buffer))
  (bury-buffer)
  (hyde/load-posts)
  nil)


(define-derived-mode hyde-markdown-mode  markdown-mode "Hyde-markdown" 
  "Markdown mode with a few extra bindings for convenience" 
  (define-key hyde-markdown-mode-map (kbd "C-c C-c") 'hyde-markdown-end-edit))
  (define-key hyde-markdown-mode-map (kbd "C-c C-v") 'markdown-preview)

(provide 'hyde-md)



hyde-git.el
;;; hyde-git.el
;; Copyright (C) 2004 Noufal Ibrahim <noufal at nibrahim.net.in>
;;
;; This program is not part of Gnu Emacs
;;
;; hyde-git.el is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published
;; by the Free Software Foundation; either version 3 of the License,
;; or (at your option) any later version.
;;
;; This program is distributed in the hope that it will be useful, but
;; WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
;; General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with this program; if not, write to the Free Software
;; Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
;; 02111-1307, USA.

(defcustom hyde/git/remote
  "origin"
  "The remote which should be pushed to"
  :group 'hyde)

(defcustom hyde/git/remote-branch
  "master"
  "The name of the branch on the remote which should be pushed to"
  :group 'hyde)

(defun hyde/git/uncommittedp (repo file)
  "Returns true if there are uncommitted changes for the current file"
  (let (
	(cmd (format "cd '%s' && git diff-files --quiet '%s' > /dev/null" repo file))
	)
    (= (shell-command cmd) 1)))

(defun hyde/git/unpushedp (repo file)
  "Returns true if there are unpushed changes for the current file"
  (let (
	(cmd 
	 (format "cd '%s' && git log --exit-code %s/%s..HEAD '%s' > /dev/null" repo hyde/git/remote hyde/git/remote-branch file)
	 ))
    (= (shell-command cmd) 1)))

(defun hyde/git/pushedp (repo file)
  "Returns true if there are no uncommitted changes in the file"
  (not (hyde/git/uncommittedp repo file)))

(defun hyde/git/add (repo file)
  "Adds the given file to the repository"
  (let ((cmd (format "cd '%s' && git add '%s' > /dev/null" repo file)))
    (shell-command cmd)))

(defun hyde/git/commit (repo file commit-message)
  "Commits the given file to the repository"
  (let ((cmd (format "cd '%s' && git commit -m '%s' '%s' > /dev/null" repo commit-message file)))
    (shell-command cmd)))

(defun hyde/git/push (repo)
  "Pushes the repository"
  (let ((cmd (format "cd '%s' && git push %s %s > /dev/null" repo hyde/git/remote hyde/git/remote-branch)))
    (shell-command cmd)))

(defun hyde/git/rename (base from to)
  "Rename the file in BASE from FROM to TO"
  (let ((cmd (format "cd '%s' && git mv '%s' '%s' > /dev/null" base from to)))
    (shell-command cmd)))


(provide 'hyde-git)


hyde.el
;;; hyde.el
;; Copyright (C) 2004 Noufal Ibrahim <noufal at nibrahim.net.in>
;;
;; This program is not part of Gnu Emacs
;;
;; hyde.el is free software; you can redistribute it and/or modify it
;; under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 3 of the License, or
;; (at your option) any later version.
;;
;; This program is distributed in the hope that it will be useful, but
;; WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
;; General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with this program; if not, write to the Free Software
;; Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
;; 02111-1307, USA.

;; Requirements
(require 'hyde-git)
(require 'hyde-md)

;; Constants for internal use
(defconst hyde/hyde-version "0.2" 
  "Hyde version")

;; Internal customisable variables
(defcustom hyde-mode-hook nil
  "Hook called by \"hyde-mode\""
  :type 'hook
  :group 'hyde)

(defcustom hyde-home 
  "/home/noufal/blog"
  "Default blog directory"
  :type 'directory
  :group 'hyde)

(defcustom hyde-deploy-dir
  "_site"
  "Directory which needs to be deployed"
  :type 'string
  :group 'hyde)


(defcustom hyde-posts-dir 
  "_posts"
  "Directory which contains the list of posts"
  :type 'string
  :group 'hyde)

(defcustom hyde-drafts-dir
  "_drafts"
  "Directory which contains post drafts"
  :type 'string
  :group 'hyde)

(defcustom hyde/hyde-list-posts-command 
  "/bin/ls -1tr "
  "Command to list the posts"
  :type 'string
  :group 'hyde)

(defcustom hyde/jekyll-command
  "jekyll"
  "Command to run jekyll to create the blog"
  :type 'string
  :group 'hyde)

(defcustom hyde/deploy-command
  "rsync -vr _site/* nkv@ssh.hcoop.net:/afs/hcoop.net/user/n/nk/nkv/public_html/nibrahim.net.in/"
  "Command used to deploy the site to the actual server"
  :type 'string
  :group 'hyde)

;; Faces and font-locking
(defface hyde-header-face
  '(
    (((type tty) (class color)) (:foreground "blue" :background "gray"))
    (((type graphic) (class color)) (:foreground "blue" :background "gray"))
    (t (:foreground "blue" :background "gray"))
    )
  "Face for a hyde header"
  :group 'hyde)

(defface hyde-committed-face
  '(
    (((type tty) (class color)) (:foreground "blue"))
    (((type graphic) (class color)) (:foreground "blue"))
    (t (:foreground "blue"))
    )
  "Face for a file that has been committed"
  :group 'hyde)


(defface hyde-modified-face
  '(
    (((type tty) (class color)) (:foreground "red"))
    (((type graphic) (class color)) (:foreground "red"))
    (t (:foreground "red"))
    )
  "Face for a file that has been modified but not committed"
  :group 'hyde)

(defface hyde-unsaved-face
  '(
    (((type tty) (class color)) (:foreground "black" :background "red"))
    (((type graphic) (class color)) (:foreground "black" :background "red"))
    (t (:foreground "black" :background "red"))
    )
  "Face for a file that has been modified but not even saved"
  :group 'hyde)

(defface hyde-pushed-face
  '(
    (((type tty) (class color)) (:foreground "green"))
    (((type graphic) (class color)) (:foreground "green"))
    (t (:foreground "green"))
    )
  "Face for a file that has been pushed to the remote repo"
:group 'hyde)

(defvar hyde-header-face 'hyde-header-face "Face for a hyde header")
(defvar hyde-committed-face 'hyde-committed-face)
(defvar hyde-modified-face 'hyde-modified-face)
(defvar hyde-unsaved-face  'hyde-unsaved-face )
(defvar hyde-pushed-face 'hyde-pushed-face)

(defconst hyde-font-lock-keywords
  (list
   '("^::.*" . hyde-header-face)
   '("^C :.*" . hyde-committed-face)
   '("^M :.*" . hyde-modified-face)
   '("^E :.*" . hyde-unsaved-face)
   '("^\\. :.*" . hyde-pushed-face)
   )
  "Font lock keywords for Hyde mode")

;; Version control abstraction
(defalias 'hyde/vc-uncommittedp 'hyde/git/uncommittedp "Command to check whether a file has uncommitted changes")
(defalias 'hyde/vc-unpushedp 'hyde/git/unpushedp "Command to check whether a file has unpushed changes")
(defalias 'hyde/vc-pushedp  'hyde/git/pushedp "Command to check whether a file has pushed changes")
(defalias 'hyde/vc-add  'hyde/git/add "Command to add a file to the DVCS")
(defalias 'hyde/vc-commit  'hyde/git/commit "Command to add a file to the DVCS")
(defalias 'hyde/vc-push  'hyde/git/push "Command to push the repository")
(defalias 'hyde/vc-rename  'hyde/git/rename "Command to push the repository")

(defun hyde/hyde-file-local-uncommitted-changed (dir file)
  "Return whether the given file in the given dir is uncommitted"
  (hyde/vc-uncommittedp (concat hyde-home "/" dir) file))

(defun hyde/hyde-file-committed-not-pushed (dir file)
  "Return whether the given file in the given dir is unpushed"
  (hyde/vc-unpushedp (concat hyde-home "/" dir) file))

(defun hyde/hyde-file-committed-pushed (dir file)
  "Return whether the given file in the given dir is pushed"
  (hyde/vc-pushedp (concat hyde-home "/" dir) file))

(defun hyde/hyde-add-file (file)
  "Stages the given file for commit."
  (hyde/vc-add (concat hyde-home "/" hyde-posts-dir) file))

(defun hyde/hyde-rename-file (from to)
  "Renames the given file from to to"
  (hyde/vc-rename hyde-home from to))

(defun hyde/hyde-commit-post (pos commit-message)
  "Commits the changes in the repository"
  (interactive "d\nMCommit message : ")
  (let (
	(post-file-name (nth 
			 1
			 (split-string (strip-string (thing-at-point 'line)) " : ")))
	(dir (get-text-property pos 'dir)))
    (hyde/vc-commit (concat hyde-home "/" dir) post-file-name commit-message)
    (hyde/load-posts)))

(defun hyde/hyde-push ()
  "Publishes the changes to the remote repository"
  (interactive)
  (hyde/vc-push hyde-home)
  (hyde/load-posts))

(defun hyde/run-jekyll ()
  "Runs jekyll on the directory"
  (interactive)
  (shell-command (format "cd %s && %s" hyde-home hyde/jekyll-command)))

(defun hyde/deploy ()
  "Deploys the generated website (should be run after hyde/run-jekyll"
  (interactive)
  (shell-command (format "cd %s && %s" hyde-home hyde/deploy-command)))
  
  
;; Utility functions
(defun hyde/hyde-file-local-unsaved-changed (dir file)
  "Returns true if and only if the given file contains unsaved changes"
    (let (
	(buffer (get-file-buffer file))
	)
    (if buffer
	(buffer-modified-p buffer)
      nil)))

(defun strip-string (str)
  "Returns STR with all trailing whitespaces gone"
  (replace-regexp-in-string "\n$" "" str))

(defun hyde/file-status (dir file)
  "Returns an letter indicating the status of the file as far as
hyde is concerned

Committed means that the changes have been committed into your DVCS
Pushed out means that they have been pushed to a safe remote repo (github, bitbucket etc.)

Status indicators are as follows:

. Committed and pushed
C Committed but not yet pushed
M Local saved changes (uncommitted)
E Local unsaved changes"
  (or 
   (and (hyde/hyde-file-local-unsaved-changed dir file) "E")
   (and (hyde/hyde-file-local-uncommitted-changed dir file) "M")
   (and (hyde/hyde-file-committed-not-pushed dir file) "C")
   (and (hyde/hyde-file-committed-pushed dir file) ".")))


(defun hyde/list-format-posts (dir)
  "Gets the lists of posts from the given directory, formats them
properly and returns them so that they can be presented to the
user"
  (let (
	(posts (split-string (strip-string (shell-command-to-string
					    (concat "cd " hyde-home "/" dir " ; " hyde/hyde-list-posts-command ))))))
    (map 'list (lambda (f) (format "%s : %s" (hyde/file-status dir f) f)) posts)))

(defun hyde/promote-to-post (pos)
  "Promotes the post under the cursor from a draft to a post"
  (interactive "d")
  (let (
	(post-file-name (nth 
			 1
			 (split-string (strip-string (thing-at-point 'line)) " : ")))
	(dir (get-text-property pos 'dir)))
    (if (equal dir hyde-drafts-dir)
	(hyde/hyde-rename-file (concat dir "/" post-file-name)
			       (concat hyde-posts-dir "/" post-file-name)))
    (hyde/load-posts)))


(defun hyde/open-post-maybe (pos)
  "Opens the post under cursor in the editor"
  (interactive "d")
  (let (
	(post-file-name (nth 
			 1
			 (split-string (strip-string (thing-at-point 'line)) " : ")))
	(dir (get-text-property pos 'dir)))
    (find-file 
     (strip-string (concat hyde-home "/" dir "/" post-file-name)))
    (hyde-markdown-mode)))

(defun hyde/new-post (title)
  "Creates a new post"
  (interactive "MEnter post title: ")
  (let ((post-file-name (format "%s/%s/%s.markdown" 
				hyde-home hyde-drafts-dir (concat 
							  (format-time-string "%Y-%m-%d-")
							  (downcase (replace-regexp-in-string " " "_" title))))))

    (find-file post-file-name)
    (insert "---\n")
    (insert "layout: post\n")
    (insert (format "title: \"%s\"\n" title))
    (insert "---\n\n")
    (save-buffer)
    (hyde/hyde-add-file post-file-name)
    (hyde-markdown-mode)))

(defun hyde/quit ()
  "Quits hyde"
  (interactive)
  (kill-buffer (current-buffer)))


;; Keymaps
(defvar hyde-mode-map
  (let 
      ((hyde-mode-map (make-sparse-keymap)))
    (define-key hyde-mode-map (kbd "n") 'hyde/new-post)
    (define-key hyde-mode-map (kbd "g") 'hyde/load-posts)
    (define-key hyde-mode-map (kbd "c") 'hyde/hyde-commit-post)
    (define-key hyde-mode-map (kbd "P") 'hyde/hyde-push)
    (define-key hyde-mode-map (kbd "j") 'hyde/run-jekyll)
    (define-key hyde-mode-map (kbd "d") 'hyde/deploy)
    (define-key hyde-mode-map (kbd "p") 'hyde/promote-to-post)
    (define-key hyde-mode-map (kbd "q") 'hyde/quit)
    (define-key hyde-mode-map (kbd "RET") 'hyde/open-post-maybe)
    hyde-mode-map)
  "Keymap for Hyde")


(defun hyde/load-posts ()
  "Load up the posts and present them to the user"
  (interactive)
  ;; Clear the buffer
  (toggle-read-only -1)
  (delete-region (point-min) (point-max))
  ;; Insert headers
  (insert ":: Editing blog at:" hyde-home "\n")
  (insert ":: Posts\n")
  ;; Insert posts from posts directory
  (let 
      ((posts (hyde/list-format-posts hyde-posts-dir)))
    (dolist (post posts)
      (progn
	(save-excursion
	  (insert (concat post "\n")))
	(put-text-property (point) (+ (point) (length post)) 'dir hyde-posts-dir)
	(forward-line))))
    (insert "\n:: Drafts\n")
    (let 
	((posts (hyde/list-format-posts hyde-drafts-dir)))
      (dolist (post posts)
	(progn
	  (save-excursion
	    (insert (concat post "\n")))
	  (put-text-property (point) (+ (point) (length post)) 'dir hyde-drafts-dir)
	  (forward-line))))
    ;; Insert footer
    (insert (concat "\n\n:: Hyde version " hyde/hyde-version "\n"))
    (insert "Key:\n-----\n . Committed and pushed\n C Committed but not yet pushed\n M Local saved changes (uncommitted)\n E Local unsaved changes\n")
    (toggle-read-only 1))

(defun hyde/read-config ()
  "Loads up the config file to set the blog deployment and other information"
  (let (
	(config-file (concat hyde-home "/.hyde.el"))
	)
    (message (format "Loading %s" config-file))
    (load-file config-file)
    ))
  

(defun hyde/hyde-mode ()
  "The Hyde major mode to edit Jekyll posts."
  (kill-all-local-variables)
  (dolist (x '(hyde-home
	       hyde-deploy-dir
	       hyde-posts-dir
	       hyde-drafts-dir
	       hyde/hyde-list-posts-command
	       hyde/jekyll-command
	       hyde/deploy-command
	       hyde/git/remote
	       hyde/git/remote-branch))
    (make-variable-buffer-local x))
  (use-local-map hyde-mode-map)
  (set (make-local-variable 'font-lock-defaults) '(hyde-font-lock-keywords))
  (setq major-mode 'hyde/hyde-mode
	mode-name "Hyde")
  (hyde/read-config)
  (hyde/load-posts)
  (hl-line-mode t)
  (run-hooks hyde-mode-hook))


;; Entry point
(defun hyde (hyde-home)
  "Enters hyde mode"
  (interactive "DBlog : ")
  (let (
	(hyde-buffer (concat "*Hyde : " hyde-home "*"))
	)
    (switch-to-buffer (get-buffer-create hyde-buffer)))
  (hyde/hyde-mode))

(provide 'hyde)
